specs/Proposals/0040-split-states.md

```diff
@@ -0,0 +1,381 @@
+```
+returns the `shard_id`s of the shards in the current epoch.
+
+#### `get_shards_next_epoch`
+```rust
+pub fn get_shards_next_epoch() -> (Vec<ShardId>, HashMap<ShardId, ShardId>)
+```
+returns the `shard_id`s of the shards in the next epoch and a map from those shards to their parent shards.
+
+## Build New States
+The following method in `Client` will be added or modified to split a shard's current state into multiple states.
+
+### `split_shards`
+```rust
+pub fn split_shards(me: &Option<AccountId>, sync_hash: CryptoHash, shard_id: ShardId)
+```
+builds states for the new shards that the shard `shard_id` will be split to.
+After this function is finished, the states for the new shards should be ready in `ShardTries` to be accessed.
+
+### `run_catchup`
+```rust
+pub fn run_catchup(...) {
+    ...
+    match state_sync.run(
+        ...
+    )? {
+        StateSyncResult::Unchanged => {}
+        StateSyncResult::Changed(fetch_block) => {...}
+        StateSyncResult::Completed => {
+            // build states for new shards if shards will change and we will track some of the new shards
+            if self.runtime_adapter.will_shards_change_next_epoch(epoch_id) {
+                let mut parent_shards = HashSet::new();
+                let (new_shards, mapping_to_parent_shards) = self.runtime_adapter.get_shards_next_epoch(epoch_id);
+                for shard_id in new_shards {
+                    if self.runtime_adapter.will_care_about_shard(None, &sync_hash, shard_id, true) {
+                	parent_shards.insert(mapping_to_parent_shards.get(shard_id)?);
+                    }
+                }
+                for shard_id in parent_shards {
+                    self.split_shards(me, &sync_hash, shard_id);
+                }
+            }
+           ...
+       }
+   }
+   ...
+}
+```
+
+## Update States
+### `split_state_changes`
+```rust
+split_state_changes(shard_id: ShardId, state_changes: &Vec<RawStateChangesWithTrieKey>) -> HashMap<ShardId, Vec<RawStateChangesWithTrieKey>>
+```
+splits state changes to be made to a current shard to changes that should be applid to the new shards. Note that this function call can take a long time. To avoid blocking the client actor from processing and producing blocks for the current epoch, it should be called from a separate thread. Unfortunately, as of now, catching up states and catching up blocks are both run in client actor. They should be moved to a separate actor. However, that can be a separate project, although this NEP will depend on that project. In fact, the issue has already been discussed in [#3201](https://github.com/near/nearcore/issues/3201).
+
+### `apply_chunks`
+`apply_chunks` will be modified so that states of the new shards will be updated when processing chunks.
+ In `apply_chunks`, after processing each chunk, the state changes in `apply_results` are sorted into changes to new shards.
+At the end, we apply these changes to the new shards.
```

Comment:
Ideally this should also be done in a different thread, but for this specific upgrade I think it is fine